modules/gapi/test/util/variant_tests.cpp

```diff
@@ -486,4 +500,209 @@ TEST(Variant, EXT_IndexOf)
     static_assert(6u == V::index_of<MyClass>(), "Index is incorrect");
 }
 
+namespace test_validation
+{
+struct MyType
+{
+    friend std::ostream& operator<<(std::ostream& out, const MyType& src)
+    {
+        return out << "MyType"; (void) src;
+    }
+};
+class MyClass
+{
+    friend std::ostream& operator<<(std::ostream& out, const MyClass& src)
+    {
+        return out << "MyClass"; (void) src;
+    }
+};
+
+struct MyParamDynamicVisitor : cv::util::static_indexed_visitor<bool, MyParamDynamicVisitor>
```

Comment:
Is this `bool` mandatory? can't it be inferred via CRTP?

Reply:
since the `static_indexed_visitor` is already using CRTP

Reply:
1) From my perspective it breaks erroneous-prone criteria:

It is possible writing  by mistake
```
class MyClass : public static_visitor<MyClass> {
bool operator()(int)
{/*1000 lines of code*/}
void operator()(float)
{/*1000 lines of code*/}
int operator() (char)
}
```
but with explicit type it less possible


AND **the more strongest** reason

 2. to deduct return_type in CRTP you must use the same construction as in visit()  `decltype(visit_impl(get<0>(var)))` where `visit_impl` is upper-level hierarchy wrapper class. But 

`decltype(visit(get<0>(var)))`  <---- where can i get `var` variable here?

Look 

```
somewhere in CRTP...

template<class T>
auto operator() (size_t index, const T& val)  -> decltype(Impl::visit(declval(T))) <---!
```
declval(T) - is useless here, because  it's declaration and doesn't define nothing because nothing to instantiate